Python/Product/PythonTools/PythonTools/LanguageServerClient/PythonLanguageClient.cs

```diff
@@ -173,9 +173,78 @@ internal sealed class PythonLanguageClient : ILanguageClient, ILanguageClientCus
+            foreach (var item in args.requestParams.items) {
+                // Find the matching context for the item
+                var context = _clientContexts.Find(c => item.scopeUri != null && PathUtils.IsSamePath(c.RootPath, item.scopeUri.LocalPath));
+                if (context == null)
+                    continue;
+
+                if (context is PythonLanguageClientContextProject) {
+                    // Project interactions are main thread only.
+                    await JoinableTaskContext.Factory.SwitchToMainThreadAsync();
+                }
+
+                object settings = null;
+                if (item.section == "python") {
+                    Debug.Assert(_analysisOptions != null);
+
+                    var extraPaths = UserSettings.GetStringSetting(
+                        PythonConstants.ExtraPathsSetting, null, Site, PythonWorkspaceContextProvider.Workspace, out _)?.Split(';')
+                        ?? _analysisOptions.ExtraPaths;
+
+                    var stubPath = UserSettings.GetStringSetting(
+                        PythonConstants.StubPathSetting, null, Site, PythonWorkspaceContextProvider.Workspace, out _)
+                        ?? _analysisOptions.StubPath;
+
+                    var typeCheckingMode = UserSettings.GetStringSetting(
+                        PythonConstants.TypeCheckingModeSetting, null, Site, PythonWorkspaceContextProvider.Workspace, out _)
+                        ?? _analysisOptions.TypeCheckingMode;
+
+                    var ver3 = new Version(3, 0);
+                    if (context.InterpreterConfiguration.Version < ver3) {
+                        MessageBox.ShowWarningMessage(Site, Strings.WarningPython2NotSupported);
+                    }
+
+                    settings = new LanguageServerSettings.PythonSettings {
+                        pythonPath = context.InterpreterConfiguration.InterpreterPath,
+                        venvPath = string.Empty,
+                        analysis = new LanguageServerSettings.PythonSettings.PythonAnalysisSettings {
+                            logLevel = _analysisOptions.LogLevel,
+                            autoSearchPaths = _analysisOptions.AutoSearchPaths,
+                            diagnosticMode = _analysisOptions.DiagnosticMode,
+                            extraPaths = extraPaths,
+                            stubPath = stubPath,
+                            typeshedPaths = _analysisOptions.TypeshedPaths,
+                            typeCheckingMode = typeCheckingMode,
+                            useLibraryCodeForTypes = true,
+                            completeFunctionParens = _advancedEditorOptions.CompleteFunctionParens,
+                            autoImportCompletions = _advancedEditorOptions.AutoImportCompletions
+                        }
+                    };
+                }
+
+                // Copy settings in order
+                var resultArray = new object[result.Length + 1];
```

Comment:
I'm going to change this to use a list and then convert the list to an array at the end.